ssg/templates.py

```diff
@@ -120,66 +132,26 @@ def __init__(
                 output_dir = self.remediations_dir
             dir_ = os.path.join(output_dir, lang_dir)
             self.output_dirs[lang_name] = dir_
-        # scan directory structure and dynamically create list of templates
+        # Scan directory structure and dynamically create list of templates
         for item in sorted(os.listdir(self.templates_dir)):
-            maybe_template = Template(templates_dir, item)
-            if maybe_template.looks_like_template():
-                maybe_template.load()
-                templates[item] = maybe_template
+            maybe_template = Template.load_template(templates_dir, item)
+            if maybe_template is not None:
+                self.templates[item] = maybe_template
         self.product_cpes = ProductCPEs()
         self.product_cpes.load_cpes_from_directory_tree(cpe_items_dir, self.env_yaml)
 
-    def build_lang_file(
-            self, rule_id, template_name, template_vars, lang, local_env_yaml):
-        """
-        Builds and returns templated content for a given rule for a given
-        language; does not write the output to disk.
-        """
-        if lang not in templates[template_name].langs:
-            return None
-
-        template_file_name = lang.name + ".template"
-        template_file_path = os.path.join(self.templates_dir, template_name, template_file_name)
-        template_parameters = templates[template_name].preprocess(template_vars, lang.name)
-        jinja_dict = ssg.utils.merge_dicts(local_env_yaml, template_parameters)
-        filled_template = ssg.jinja.process_file_with_macros(
-            template_file_path, jinja_dict)
-
-        return filled_template
-
-    def build_lang(
-            self, rule_id, template_name, template_vars, lang, local_env_yaml, platforms=None):
-        """
-        Builds templated content for a given rule for a given language.
-        Writes the output to the correct build directories.
-        """
-        if lang not in templates[template_name].langs or lang.name == "sce-bash":
-            return
-
-        filled_template = self.build_lang_file(rule_id, template_name,
-                                               template_vars, lang,
-                                               local_env_yaml)
-
-        ext = lang.file_extension
-        output_file_name = rule_id + ext
-        output_filepath = os.path.join(
-            self.output_dirs[lang.name], output_file_name)
-
-        with open(output_filepath, "w") as f:
-            f.write(filled_template)
-
-    def get_langs_to_generate(self, rule):
+    def get_langs_to_generate(self, id_, template):
```

Comment:
What was the motive of moving from rule properties to concrete fields?

- On the good side, arguments of the method got narrowed down more, which is always good, but
-  the method itself references rules in in messages and in its docstring.

The method is used in rule context only, so shouldn't the new arguments be renamed to `rule_id` and `rule_template` then?

Reply:
Everything is even simpler: the list of languages is based on the properties of a template and template only. Even the id is used only to show an error message (very controversial decision, but I left it there for to minimize the amount of changes to review). So, the context of the method is template. And even the method itself could be static or a module function.

Reply:
Scratch that. After giving it another 15 seconds I realized that even the message is bogus. If the function does not depend on the properties of a rule there is no sense in mentioning its ID.

Reply:
Err. Actually no, it's a bit more complicated. The template is not the template but a subsection of rule settings. Oh, well.